File: backend/models/access.py
"""Access and billing API models; mirrored in frontend/src/lib/access.ts."""
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


class UserView(BaseModel):
    user_id: str
    name: str
    email: str
    picture: str = ""
    has_access: bool
    access_until: datetime | None = None
    access_mode: Literal["live", "test", "admin"] | None = None


class AuthState(BaseModel):
    user: UserView | None = None


class SessionExchange(BaseModel):
    session_id: str = Field(min_length=8, max_length=512)


class Message(BaseModel):
    message: str


class AdminUnlock(BaseModel):
    password: str = Field(min_length=1, max_length=200)


class Plan(BaseModel):
    amount: int = 34900
    currency: str = "INR"
    months: int = 6
    enabled: bool
    mode: Literal["disabled", "test", "live"]


class OrderView(BaseModel):
    order_id: str
    key_id: str
    amount: int
    currency: str
    mode: Literal["test", "live"]


class PaymentVerification(BaseModel):
    razorpay_order_id: str = Field(pattern=r"^order_[A-Za-z0-9]+$", max_length=100)
    razorpay_payment_id: str = Field(pattern=r"^pay_[A-Za-z0-9]+$", max_length=100)
    razorpay_signature: str = Field(pattern=r"^[a-f0-9]{64}$")

File: frontend/src/lib/access.ts
// Mirrors backend/models/access.py. Auth tokens never enter JavaScript.
import { useQuery } from "@tanstack/react-query";
import { apiGet } from "./api";
export interface UserView {
  user_id: string; name: string; email: string; picture: string;
  has_access: boolean; access_until: string | null; access_mode: "live" | "test" | "admin" | null;
}
export interface AuthState { user: UserView | null }
export interface SessionExchange { session_id: string }
export interface Message { message: string }
export interface AdminUnlock { password: string }
export interface Plan { amount: number; currency: string; months: number; enabled: boolean; mode: "disabled" | "test" | "live" }
export interface OrderView { order_id: string; key_id: string; amount: number; currency: string; mode: "test" | "live" }
export interface PaymentVerification { razorpay_order_id: string; razorpay_payment_id: string; razorpay_signature: string }
export function useAuth() {
  return useQuery({ queryKey: ["auth"], queryFn: () => apiGet<AuthState>("/auth/me"),
    retry: false, staleTime: 0, refetchInterval: 30000, refetchOnWindowFocus: "always" });
}
export function googleLoginUrl() {
  // REMINDER: DO NOT HARDCODE THE URL, OR ADD ANY FALLBACKS OR REDIRECT URLS, THIS BREAKS THE AUTH
  const redirect = window.location.origin + "/account";
  return `https://auth.emergentagent.com/?redirect=${encodeURIComponent(redirect)}`;
}
export function accessDate(value: string | null) {
  return value ? new Intl.DateTimeFormat("hi-IN", { dateStyle: "long", timeStyle: "short", timeZone: "Asia/Kolkata" }).format(new Date(value)) : "—";
}